semantics fixes for new-roll-engine

rsek opened the pull request:
- [ ]
- [ ] Update CHANGELOG.md

ben reviewed (commented):
I'm liking most of the renaming here, lots of this is definitely better. It looks like there are still bugs hiding, definitely around the outcome values (see line comments). And I would have told you to not worry about the handlebars helpers and `chatrollhelpers.ts`, I'm hoping to delete those files sometime real soon.

Also I found some visual bugs:

This one is missing the "= 7" result text, and the outcome text is wrong, and the inset white area is missing some vertical whitespace from before.

![CleanShot 2022-08-20 at 19 54 50](https://user-images.githubusercontent.com/39902/185773382-658eaa55-2fad-4b5d-b580-0e895d91b79b.jpg)

This one shows that expanding the new unsettled field sets makes it so you have to scroll the dialog to see the buttons again. Also there's a translation bug.

![CleanShot 2022-08-20 at 19 54 08](https://user-images.githubusercontent.com/39902/185773389-54f161b5-1a6b-4267-a78c-adf065e274f8.jpg)

rsek commented on src/module/rolls/chat-message.ts:
```diff
@@ -7,21 +7,37 @@ import { SFMoveDataProperties } from '../item/itemtypes'
 import { renderRollGraphic } from './roll-graphic'
 
-function rollResult(a: number, c1: number, c2: number): ROLL_OUTCOME {
+function resolveIronswornRoll(a: number, c1: number, c2: number): ROLL_OUTCOME {
```
my first instinct there would be`resolveRollOutcome` (if only to keep to the convention of functions being verbs), but fair enough on "resolve" not making sense here 🤔 what do you think of `computeRollOutcome`?

ben replied:
👍🏼

rsek commented on src/module/rolls/chat-message.ts:
```diff
@@ -7,21 +7,37 @@ import { SFMoveDataProperties } from '../item/itemtypes'
-  let key = 'WeakHit'
-  if (outcome === ROLL_OUTCOME.MISS) {
-    key = match ? 'Complication' : 'Miss'
+function momentumBurnIsValid(
+  rawOutcome: ROLL_OUTCOME | undefined,
+  momentumOutcome: ROLL_OUTCOME
+) {
```
oh, yeah, that makes sense. another possibility is `momentumIsBurnable`, but that speaks more to the purpose of the check rather than what the function is actually _doing_. so i reckon `momentumBurnWouldUpgrade` is better.

aside: this is one reason why i kinda like a numeric enum for roll outcomes -- makes comparing them dead simple.

rsek commented on src/module/rolls/preroll-dialog.ts:
```diff
@@ -70,6 +70,7 @@ function prerollOptionsWithFormData(
   const opts = cloneDeep(base)
 
   const valMap: Record<string, string> = form
+    // "in particular the element cannot be disabled and must contain a name attribute"
```
ah, yeah, had to look up something in the jquery docs to grok why some nearby code had to be done a certain way. just forgot to clean it up after.

rsek commented on src/module/rolls/roll.ts:
```diff
@@ -10,10 +10,16 @@ import { getFoundryMoveByDfId } from '../dataforged'
 import { IronswornItem } from '../item/item'
 import { IronswornRollChatMessage } from './chat-message'
 
+export const ROLL_SCORE_MAX = 10
```
technically, it's also the progress score maximum. there isn't a 'canon' term that encompasses both the progress score *and* the action score, but 'score' is generally what ive called it in other implementations. `SCORE_MAX` alone seemed a little ambiguous, hence the prefix of `ROLL_` but maybe it'd be fine with little explanatory tsdoc annotation? alternatively, could namespace it alongside a bunch of other ironsworn roll-related constants 🤔

rsek commented on src/module/rolls/roll.ts:
```diff
@@ -291,25 +301,25 @@ export class IronswornRoll {
   }
 
-  get preAdjustmentOutcome(): SourcedValue<ROLL_OUTCOME> | undefined {
+  get rawOutcome(): SourcedValue<ROLL_OUTCOME> | undefined {
```
hmm. is there an outcome that's *more* raw than this one? possible i just misunderstood the sequence.
maybe a better question to ask is... would it make sense to standardize some of the terms there based on what part of the roll process they happen in?

ben replied:
Gosh, we have to give them names? I think the states for any of the terms in the Action Equation™ are:

1. Unknown yet
1. Pre-determined
1. Rolled or calculated
1. Replaced after the roll

For display purposes, we can treat states 2 and 3 mostly identically (aside from recording where the value came from), but there are still three cases to distinguish. If a roll had its outcome changed by burning momentum, we'd like to be able to store that state, and display it nicely.

"Unknown" seems like a good fit for state 1. For the others:

* 2-3: "natural" or "original"
* 4: "overridden" or "replaced"

rsek commented on system/templates/rolls/preroll-dialog.hbs:
```diff
@@ -22,63 +26,72 @@ roll {{Method}} of {{join Using delimiter=', '}} {{!-- TODO: i18n --}}
   {{/if}}
   {{/if}}
-
+{{!-- FIXME --}}
```
lol, no idea.

rsek commented on system/templates/rolls/roll-graphic.hbs:
```diff
@@ -22,7 +24,7 @@
       <strong>
         {{#if actionTotalCapped}}
-        <abbr title="{{localize 'IRONSWORN.CappedAt10'}}">{{actionTotal}}</abbr>
+        <span title="{{localize 'IRONSWORN.CappedAt10'}}">{{actionTotal}}</span>
```
oh, i can cook an identical-looking css class up for that in a jiffy :)

rsek commented on src/module/rolls/roll-graphic.ts:
```diff
@@ -68,21 +71,13 @@ export async function renderRollGraphic(
-        outcomeKeys[referenceRoll.preAdjustmentOutcome.value]
-      ),
+      ...referenceRoll.rawOutcome,
+      value: game.i18n.localize(referenceRoll.rawOutcome.value),
```
we discussed this some in the discord, but to include an overview of that for posterity: yeah, i think my real gripe here is the noise added by repeated string transforms (which i get paranoid about, having been burned by them before). but there's other ways of managing that, like by extracting the transforms as function, and possibly getting a numeric enum involved. i'll noodle on that more and see if i come up with anything clever.

rsek commented on src/module/rolls/roll.ts:
```diff
@@ -79,25 +85,29 @@ export class IronswornRoll {
   }
-
-  static action(stat: string, value: number, adds?: number): IronswornRoll {
+  // TODO: a way to add arbitrary values for certain assets (Looper) and moves (Develop your Relationship)?
```
cool - i'll add a comment to that effect.

knowing the intended use, though -- i think there's utility in this being able to set arbitrary number values unlinked to a stat/condition meter in place of a reference to a PC attribute. aside from the 'canonical' outliers in Looper and Develop Your Relationship, custom macros seem like exactly the sort of place where a user is more likely to need some offbeat stat.

reckon it'd work if the `stat` argument was typed as `number | string`? then if `typeof stat === 'string'`, it's presumed to be an attribute key. if `typeof stat === 'number'`, the value is used.

slightly further afield: what do you think of including an optional argument for a move ID? then the macro can include a move reference, if desired; the chief benefit from the user PoV is that it'd include the relevant move outcome text in the chat output. heck, this could even *override* a move's normal stat/progress roll options. this would provide one avenue for people to implement alternate stat rolls w/o necessarily editing the moves themselves.

ben replied:
Hmm. Maybe we should just remove these helpers until we can think through what they're needed for. You're right, this one probably isn't as helpful as I had intended when I started.

rsek commented on src/module/rolls/chat-message.ts:
```diff
@@ -7,21 +7,37 @@ import { SFMoveDataProperties } from '../item/itemtypes'
+    key += '_match'
   }
-  return game.i18n.localize('IRONSWORN.' + key)
+  return game.i18n.localize('IRONSWORN.' + key) as RollOutcomeKey
```
yeah, i think i just got mixed up on what this function was doing, hah

ben commented on src/module/helpers/handlebars.ts:
```diff
@@ -1,34 +1,42 @@
-    .filter((x) => x)
-    .join(' ')
+  const maxRoll = d?.faces || SCORE_MAX
+  const cssClasses = new DOMTokenList()
```
I wasn't reviewing the changes to this file very closely, because it's essentially on its way out the door, but you did break move-rolling with the Ironsworn move sheet here. This is Chrome, which means it'll also be broken in the app.

```
handlebars.ts:28 Uncaught (in promise) TypeError: Illegal constructor
[Detected 1 package: system:foundry-ironsworn]
    at classesForRoll (handlebars.ts:28:22)
```

Can we just drop all the changes to this file that aren't necessary from renames?

ben commented on src/module/chat/chatrollhelpers.ts:
```diff
@@ -1,3 +1,5 @@
+// this file is basically deprecated. use cards.ts instead
```
```suggestion
// this file is basically deprecated. use rolls/chat-message.ts instead
```

ben commented on src/module/rolls/chat-message.ts:
```diff
@@ -141,8 +157,7 @@ export class IronswornRollChatMessage {
     // FIXME: oof. so, "roll" is a tricky word since in english it can function as a verb or a noun. it gets even more complicated when you introduce grammatical gender and word order.
```
If this bothers you, we can introduce a new i18n key.

```json
"RollPlusStat": "Roll +{stat}"
```

Which you can invoke using format:

```ts
game.i18n.format('IRONSWORN.RollPlusStat', {stat: plusStat})
```

ben commented on src/module/rolls/roll.ts:
```diff
@@ -27,48 +52,81 @@ export type SourcedValue<T = number> = {
-  // Exactly one of `progress` or `action` is required
-
-  // indicates this is a progress roll, no action die will be rolled
+  // TODO: candidate for being a generic type with some variable props dependent on boolean "isProgressRoll"?
```
In the interest of getting this merged and shipped, let's leave this as a problem for Future Rsek to solve. :)

ben commented on src/module/rolls/roll.ts:
```diff
@@ -27,48 +52,81 @@ export type SourcedValue<T = number> = {
 }
 
+/**
+ * This is intended as a way for (e.g.) a macro author to make use of this class without having to instantiate a complex object as an input. It's a shortcut, and we're not using it anywhere in the code. Given that use case, I don't think we need a TODO here to support passing in assets, because that's logic for a different layer — this class does rolling and storage, it doesn't know about assets.
```
Probably doesn't need to be in the code.

rsek replied:
i'll make the tsdoc annotation more concise. main reason i dumped your description in there is so i'd remember that "IronswornRoll" has that specific usage. another possibility is renaming the class... but ideally it has both an informative name *and* a tsdoc annotation, IMO.

ben commented on system/templates/rolls/roll-graphic.hbs:
```diff
@@ -1,17 +1,19 @@
-  <div class="dice-formula">
-    <p>
+<article class="dice-roll ironsworn-roll">
+  <section class="dice-formula">
```
Not sure if you've tested this, but we lost some vertical whitespace here.

Before:

![CleanShot 2022-08-21 at 06 55 04](https://user-images.githubusercontent.com/39902/185794355-51f7d6c0-4844-4941-b015-c26d330db5d4.jpg)


After:

![CleanShot 2022-08-21 at 06 52 42](https://user-images.githubusercontent.com/39902/185794309-68758b94-37aa-4989-8f5d-f6a959b42a31.jpg)

Also, the action score total is still missing, and the outcome is still mistranslated.

rsek replied:
yeah, the styling is often the absolute *last* thing i look at, haha. the semantic content of HTML is much more rigid than CSS, which is ultimately just rendering parameters. the CSS styling is definitely important, but should be informed by the semantics and not the other way around. practically speaking, most non-form HTML elements are just restylable boxes anyway, so they're roughly interchangeable as far as CSS is concern - they're about what the content *is*, rather than a vessel for applying styles *per se*.

the bad `Strong_hit` string, on the other hand, is pure goof on my part ;)

ben commented on src/module/rolls/roll-graphic.ts:
```diff
@@ -14,14 +14,16 @@ type RenderData = {
 
-  challengeDice: Partial<SourcedValue & { minmax?: string }>[]
+  challengeDice: Partial<
+    SourcedValue & { minmax?: 'min' | 'max' | undefined }
```
Nit: the `?:` here includes the `| undefined` part of the type.

rsek replied:
oh, ive been meaning to bring that up, actually - most TS best practices i've seen recommend using [strict null checks](https://www.typescriptlang.org/tsconfig#strictNullChecks). we don't currently have that flag set, but maybe it'd be a good idea?

ben replied:
Separate PR, my friend. 😀

rsek commented:
* move outcomes are now handled as a numeric outcome. this allows for much more concise code in many places
* lots of fiddling with locale strings; several new ones drawn from the R20 localization data
* added placeholder character, "?", to empty dice on preroll dialog, to help telegraph it's indeterminate state. mostly i just wanted to see how it'd look, but i kinda like it

some notes on stuff that i'd still like to fix:

## general
* make SourcedValue have a separate `label` field? it seems like its `source` property is pulling double duty as both a value used by internal checks *and* as strings, which makes me uneasy - it's harder to do any typing there, and it's tricky to provide additional annotation outside of the HB templates.
* there's enough roll-specific helper functions/constants at this point that that they should probably be their own file 😬 


## preroll dialog

<img width="534" alt="Screen Shot 2022-08-22 at 3 35 08 AM" src="https://user-images.githubusercontent.com/5354757/185901423-c61243b8-87dd-458c-8dac-a1d6178a3917.png">

* currently, anything that refreshes the roll dialog causes the roll graphic to break -- not sure what's happening there 🤔 
* we probably don't need to see "(Marc Progress)" again on the button
* ideally the checkboxes would disable the corresponding select/input when unchecked
* bunch of button labels and titles need to be tweaked due to some casing changes
  * lazy fix: `text-transform: uppercase` on all of them?

## roll output (general)
* visual indicator of pre-set value - e.g. a little lock icon?
  * could defer 'til later. in the mean time, just add a class to their `<span>` to identify them, and we can sort the styling out later?
* special styling for matches - they're at least as important as min/max!
  * i've added a "match" class to a div that holds both challenge dice for now

## generic roll output
<img width="303" alt="Screen Shot 2022-08-22 at 3 21 04 AM" src="https://user-images.githubusercontent.com/5354757/185898975-c316b19f-9be9-4755-86a0-512fde1e4523.png">
* casing on 'roll +health' can be disguised with CSS text-transform or the now slightly-more-locale-friendly handlebars helper
* we don't need a second instance of the PC name in parentheses

## move-specific roll output
<img width="312" alt="Screen Shot 2022-08-22 at 3 15 04 AM" src="https://user-images.githubusercontent.com/5354757/185897839-7e900ced-8d6c-4cdb-9eb6-3768589da222.png">

* "Face Danger (heart)" would be better rendered as "Face Danger +heart"
* we don't need two instances of "Weak hit". either we can remove the "= Weak hit" part, or the one preceding the move text
  * personally i think the outcome label is important enough that dedicating a line to it is OK. especially since match information is of some importance, and those just make the strings longer, especially in german
* keeping a consistent number of spaces/'slots' for dice terms would be better for a quick read, IMO - then everything is in the same place for every action roll. in other words, "Adds" would always be present as a single item, and it just shows "0" if no add was supplied. further breakdowns of adds -- in the rare cases they come up, like in post-roll adds -- are better conveyed through title/tooltips
* it'd be good to see the action score total in addition to the arithmetic

ben commented on src/module/chat/cards.ts:
```diff
@@ -128,11 +129,7 @@ export class IronswornChatCard {
-        [ROLL_OUTCOME.WEAK_HIT]: 'Weak_hit',
-        [ROLL_OUTCOME.MISS]: 'Miss',
-      }[hittype]
+      const i18nKey = ''
```
Somewhere along the line, momentum-burn updates got broken. I think this is the culprit for the specific breakage:

![CleanShot 2022-08-22 at 06 21 50](https://user-images.githubusercontent.com/39902/185931789-c6f9705e-0df1-413e-9e2f-af5e20ecf863.jpg)

ben commented on src/module/chat/chatrollhelpers.ts:
```diff
@@ -179,14 +161,8 @@ function calculateMoveResultText(
-      return move.Strong
-  }
+  const dfOutcomeKey = DfRollOutcome[type]
+  return move[dfOutcomeKey]
```
This change is incorrect. `move` here is an `EnhancedDataswornMove`, not a Dataforged-derived thing. Which is why Ironsworn-sheet move rolls don't display an outcome.

#422:
![CleanShot 2022-08-22 at 06 35 14](https://user-images.githubusercontent.com/39902/185934116-b15fcc7e-3c62-48db-b6da-d17e7685be11.jpg)


This patch:
![CleanShot 2022-08-22 at 06 31 31](https://user-images.githubusercontent.com/39902/185933374-c4bc7a7a-66e3-422a-9ff6-c2d9f7bc1b43.jpg)

ben commented on src/module/chat/chatrollhelpers.ts:
```diff
@@ -195,14 +171,13 @@ function calculateSFMoveResultText(
+  if (match) {
+    return parentOutcome['With a Match']?.Text ?? parentOutcome.Text
+  }
+  return parentOutcome.Text
```
Why the name change? What is this a parent or child of?

rsek replied:
"parent" in the semantic sense ("Strong hit" is parent to "Strong hit with a match"), but i reckon you're right on it being confusing -- i'll just rename it to `outcome`.

ben commented on src/module/rolls/chat-message.ts:
```diff
@@ -68,47 +68,61 @@ export function formatRollMethod(rollMethod: RollMethod, stats: string[]) {
+  outcome: ROLL_OUTCOME,
+  match: boolean
+): i18nOutcomeKey {
+  let key: i18nOutcomeKey = ROLL_OUTCOME[outcome] as keyof typeof ROLL_OUTCOME
```
Hmm. This is sneaky and clever, and it confused me for a bit. Is it a good idea to do `enum[1]`? Or is that well-defined behavior everywhere?

rsek replied:
hmm, good question. IME it works, but those have been different contexts. i can't recall reading anything either way as far as whether it's a good idea in all contexts. i'll look to see if i can determine what the consensus on the topic is.

rsek replied:
here's the TS documentation on reverse mappings in numeric enums, which includes the JS it generates for them. included is an example similar to `enum[1]`.

the JS is compiles to is basically a lookup function.

https://www.typescriptlang.org/docs/handbook/enums.html#reverse-mappings

but the docs doing it makes a strong case, it doesnt necessarily make it a good idea, either. the opinion the author presents here is fairly partisan (in favour of TS enums), but they include links to several articles by people who hold the opposing view.

https://dev.to/dvddpl/whats-the-problem-with-typescript-enums-2okj

here's my take after digging in to those: this enum is  a concise way of representing a bidirectional lookup that's usable at runtime (unlike, say, a `type` expression). there's already several unidirectional lookup functions (and miscellaneous hash objects) being used for the same purpose in this system's code, so it's not really doing anything that's not already happening.

so, i think the takeaway here is: if you think the resulting compiled JS for this enum isn't especially offensive -- and has no marked disadvantages compared to a hand-coded lookup function -- i reckon we might as well use it :)

ben commented on src/module/rolls/chat-message.ts:
```diff
@@ -260,21 +274,15 @@ export class IronswornRollChatMessage {
-        }
-      default:
-        return {}
+    if (!isUndefined(rawOutcome) && momentumOutcome > rawOutcome) {
```
I know it's not much shorter, but wouldn't it still make sense to use `momentumBurnWouldUpgrade` here? For clarity of intent if nothing else.

rsek replied:
oddly enough, originally i kept it as its own `momentumBurnWouldUpgrade`. i'm open to going back to that, but here's my rationale for its current state:

* the only code inside `momentumBurnWouldUpgrade` was `return momentumOutcome > rawOutcome`, which seemed very lonely :pensive:
* another possibility i considered was renaming `momentumBurnWouldUpgrade` to something that expressed what it *does* (comparing two different move outcomes, which might have other applications) rather than what it's *used by*

the thought experiment that clinched it for me was: if i was unaware of the numeric move outcome enum, but broadly understood the game and what "outcome" referred to, what would i make of `momentumOutcome > rawOutcome`? 

the conclusion i reached on that was: i'd probably get the jist of "this is a comparison of two move outcomes, one from momentum and 'raw' one (whatever that is), to find out if one is higher than the other". big YMMV on that, though.

on the other hand, i reckon there's some value in functions beyond the code within them - like, it also abstracts the code into a human-friendly "label" on what's happening... and it can serve as an attachment point for documentation.

how about this: i give it a pass for more annotation via comments like youve already established within other functions, breaking it down step by step. if that looks OK, that might be the best of both worlds. if it doesn't, i can put it in a function again and we'll still have the annotations.

ben commented on src/module/rolls/preroll-dialog.ts:
```diff
@@ -119,6 +122,7 @@ export class IronswornPrerollDialog extends Dialog<
 
   static get defaultOptions() {
     return mergeObject(super.defaultOptions, {
+      // FIXME: 'dialog' class may be unnecessary since the parent object will reliably have [role=dialog] selector
```
This is a carrying forward of the parent class's default options. Here's what it looks like in foundry.js:

```js
class Dialog extends Application {
  [...]
  static get defaultOptions() {
    return foundry.utils.mergeObject(super.defaultOptions, {
      template: "templates/hud/dialog.html",
      classes: ["dialog"],
      width: 400,
      jQuery: true,
    });
  }
```

`mergeObject` doesn't extend arrays, so we have to include the original value here.

ben commented on src/module/rolls/preroll-dialog.ts:
```diff
@@ -69,11 +75,14 @@ function prerollOptionsWithFormData(
+  const valMap: Record<string, number> = form
     .serializeArray()
-    .reduce((coll, { name, value }) => ({ ...coll, [name]: value }), {})
+    .reduce(
+      (coll, { name, value }) => ({ ...coll, [name]: parseInt(value, 10) }),
+      {}
+    )
```
The use of `parseInt` here breaks the parsing of the checkboxes, which prevents the pre-roll dialog from showing extra challenge dice, automatic outcomes, or preset action dice.

rsek replied:
ooh, maybe that's the source of the roll graphic breaking a dialog form change. i'll make the function a bit more sensitive to that, then

ben commented on src/module/rolls/roll.ts:
```diff
@@ -41,12 +41,45 @@ export const DIE_LOWEST_FACE = 1
+  'Strong Hit' = 2,
 }
 
+// TODO: consider differentiating `source` from a new prop called e.g. `label`; it's pulling double duty now as both as something being tested for internal logic *and* as a user-facing label. even if those are mutually exclusive categories right now, it might be risky in the long term to conflate them.
```
Is this true though? I don't see any places where it's used as anything but an annotation that will eventually make it to the UI.

rsek replied:
my thinking here is: "thing that is checked by internal logic" and "think that gets provided to the user as a label" are very different purposes with different needs, which i'd tend towards making their own keys (having been bitten by "eh, it's *pretty close* in the past" on dataforged, lol). *right now* they overlap enough that it's no big deal, but if those needs expand (e.g. by expanding localization), that might change.

i reckon part of what makes me nervous is relying on various strings being changed (or, to frame it using the "different purposes/needs" above: their purpose/needs being changed midstream). it strikes me as something that could be a bit fragile, especially with locale strings (and user-provided strings e.g. labels drawing from custom asset meter names) in the mix.

YMMV on that, though. FWIW, if i were trying to address some of the above but keep it down to 2 props, i might structure it something like:

* keep **anything** that could be user input out of `source` entirely
* avoid localized i18n strings; if *absolutely* necessary, use their key instead 
* where possible, enumerate valid `source` values with an enum (it might then be typed as a union of those enums)
* a more elaborate version might also include:
  * FVTT ids, perhaps prefixed with item type using a reliable delimiter e.g. `item:dkhqeleflq1255r`
    - include a way to reference their properties e.g. `item:dkhqeleflq1255r#data.whatever`, to allow for e.g. referencing the label of a custom asset meter

but all that actually seems like more work (and is probably less flexible than) than just adding an optional third prop, where it's safe to put any old string because it'll never be parsed.

rsek replied:
wow, that was a lot of words! anyways -- the answer to all of that might be simply: "its unlikely to cause problems, so it's good enough for now so it gets into people's hands and get feedback to iterate on". fair enough -- just want to register my concerns before it's released into the wild ;)

rsek replied:
actually: i've just run into a situation where this makes it difficult to make even minor adjustments to string composition, spotted as i was trying to restore proper casing to e.g. stat names.
<img width="215" alt="Screen Shot 2022-08-23 at 2 27 36 AM" src="https://user-images.githubusercontent.com/5354757/186123453-a701a5ab-a70e-45dc-8378-df7477f9d4c1.png">
doesn't seem optimal to have the name there twice. ideally, it would only show "(Marc Progress)" if the message was being attributed to someone other than Marc Progress.

sounds simple enough. but the `source` string being a prebuilt label, "health (Marc Progress)", means i have to write something to parse the name back out of the string... and that all assumes it's the same actor in the first place (they might just have the same name).

and god help me if i try to do that with an actor.data.name that includes any parentheses. most names don't. but there's some not-uncommon cases for that, just off the top of my head:

* in FVTT specifically, i've seen it used to denote PC ownership ("Charname (playername)") or pronouns ("Charname (he/him)") of a character in ways that FVTT's UX doesn't address (e.g.:  to indicate 'ownership' of a character in a game where everyone is a GM, or if they want them to be easily available without opening the character sheet)
* in multilingual settings in the real world, someone might provide the name they use in one languge, and then in parentheses provide the name they use in another language. in vancouver i see this pretty regularly: someone's english name is followed by their name's original rendering as hanzi characters wrapped in parentheses

ben replied:
Okay, what if instead of adding another field to `SourcedValue`, we instead add a parameter to `PreRollOptions` to indicate the "title" or "name" or "caption" for the roll? I think we both get what we want there; the `source` field stays as (mostly) hard-coded-through-i18n hints that (mostly) end up in tooltips, and we can still augment the roll title for asset tracks.

rsek replied:
yeah, `title` makes a lot of sense to me - maybe as an optional that falls back to its present value? then it's a bit easier to change incrementally

ben commented on src/styles/styles.less:
```diff
@@ -623,6 +623,16 @@ hr {
+  justify-content: center;
+  flex-direction: row;
+}
+
```
Could we do this as well? It would make the preroll dialog look nice again. 😄 

```suggestion

fieldset {
  padding: 0;
  border: none;
}
```

ben commented:
> * there's enough roll-specific helper functions/constants at this point that that they should probably be their own file 😬

You're probably right. I hate having a bunch of files called `helper.ts` or `util.ts` but it's starting to feel warranted.

> * currently, anything that refreshes the roll dialog causes the roll graphic to break -- not sure what's happening there 🤔

I found and fixed this, and pushed a commit. (You changed the class of the graphic's root element, and didn't change the code that finds and replaces it.)

> * we probably don't need to see "(Marc Progress)" again on the button

This is one area where we're probably using the `source` field too much. But — BUT — try clicking the roll button on an asset's track, and you'll see this actually adds clarity. Marc might be rolling, but it's the cat's health being rolled.

![CleanShot 2022-08-22 at 07 46 31](https://user-images.githubusercontent.com/39902/185950298-448ff483-8abc-487a-9a67-9858a4123acd.jpg)

> * ideally the checkboxes would disable the corresponding select/input when unchecked

There's a lot we can do here, but I **really** want to get this merged so we can move forward on the stuff it's blocking. It's already _better_, let's not let perfect be the enemy of good.

> * bunch of button labels and titles need to be tweaked due to some casing changes
>   * lazy fix: `text-transform: uppercase` on all of them?

TBH this is probably on-brand, especially on the buttons. There are a number of places this text ends up, so replacing the i18n like you're doing was always going to be a bit risky.

> * visual indicator of pre-set value - e.g. a little lock icon?
>   * could defer 'til later. in the mean time, just add a class to their `<span>` to identify them, and we can sort the styling out later?
> * special styling for matches - they're at least as important as min/max!
>   * i've added a "match" class to a div that holds both challenge dice for now

Remember #366, and don't get distracted by making _these_ dice look a bit fancier when we're about to replace them.

> * we don't need a second instance of the PC name in parentheses

See example above; it adds clarity when you're rolling the health of an asset.

> * "Face Danger (heart)" would be better rendered as "Face Danger +heart"

Not super pressing right now, but in general I agree.

> * we don't need two instances of "Weak hit". either we can remove the "= Weak hit" part, or the one preceding the move text
>   * personally i think the outcome label is important enough that dedicating a line to it is OK. especially since match information is of some importance, and those just make the strings longer, especially in German
> * keeping a consistent number of spaces/'slots' for dice terms would be better for a quick read, IMO - then everything is in the same place for every action roll. in other words, "Adds" would always be present as a single item, and it just shows "0" if no add was supplied. further breakdowns of adds -- in the rare cases they come up, like in post-roll adds -- are better conveyed through title/tooltips
> * it'd be good to see the action score total, preferably marked in a way that's identical to the progress score. then it's very easy to pick out the important numbers from the lineup

Save these for later; remember #366.

## Broken Momentum

It looks like momentum burning is broken. Here's a move roll from a SF sheet (which goes through `rolls/chat-message.ts`):

![CleanShot 2022-08-22 at 07 55 56](https://user-images.githubusercontent.com/39902/185952897-74866f8c-69b6-4ec6-b137-c48b8340341b.jpg)

And here's a move roll from an Ironsworn sheet (which goes through `chatrollhelpers.ts`):

![CleanShot 2022-08-22 at 07 58 05](https://user-images.githubusercontent.com/39902/185953291-fdd680dd-2891-41c7-a92c-8dd5e799766e.jpg)

ben commented on system/templates/rolls/preroll-dialog.hbs:
```diff
@@ -1,84 +1,117 @@
+<form class='preroll-dialog'>
   {{#if move}}
-  <div>{{{enrichMarkdown move.data.data.Trigger.Text}}}</div>
+    <style>
+      .roll-trigger-text { }
+    </style>
```
Is it your intent to include global styles in every dialog that's rendered?

rsek replied:
more just a convenience to keep them together - same principle as SFCs. unfortunately, i discovered that doesnt actually work except for handlebars static sites (obvious in hindsight :sweat_smile:).

i think the ideal solution would be some vite processing to extract the styles themselves - so, source handlebars files live in `src`, then get processed and dumped in system, while the contents of their style tags is included in the standard css file.

a less labor-intensive method of component-izing their specific styling would be something like: there's a preroll-dialog.less file that lives next to this one, and it's imported when the style sheet is build.

short term, i just can move them to the main styles.less file, tho. i'll include this template's name in a comment.

rsek commented:
whoops, only spotted this post now - i put a whole reply above before realing you'd already addressed it here 🤦 

the sticking point for me is name repetition. i agree that this isn't necessarily a "fix now" problem. maybe that's just a starting point so it gets out the door, and that's fine. but since the underlying `SourcedData` currently makes it very difficult to display that information only when it's useful... i have misgivings. does that have implications within the scope of this PR? i reckon that depends on how much add'l work would be created by having to revise `SourcedData` once it's established (if the problems i anticipate actually crop up, that is 😉 ).

> This is one area where we're probably using the `source` field too much. But — BUT — try clicking the roll button on an asset's track, and you'll see this actually adds clarity. Marc might be rolling, but it's the cat's health being rolled.

oh yeah, i totally agree that that the source information needs to be provided in the case of assets... but i'm not sure the message's primary heading is the best place to do it. it takes up a lot of space in the title, putting add'l pressure on some localizations.

not *every* roll will suffer for this -- progress rolls are fine, for instance. but, i'll put it this way: if we assume that, say, 65% of all rolls are coming directly from the PC's stats/meters rather than progress or a specific asset's unique condition meter, that means that 65% of roll output will have a bunch of extraneous text in the header.

and if the `SourcedData` inherently has difficulties with providing concise text output (not just aesthetics - improving the signal:noise ratio is generally an accessibility gain) for that majority of roll output (by e.g. making the source rendering conditional on its relevance)... i'm not sure how well it's serving its purpose.

speaking of accessibility - ideally, any source labels would have some annotation describing just what the parenthetical element is describing. that'd mean wrapping the parentheses in an html element to provide it with e.g. `aria-label` (and, post v10, a tooltip so ppl not using screen readers can benefit from that extra code, too). the only way to target the parenthetical asset name would be to embed some html... so then we're putting html in `source` strings. or, it can wrap the entire "stat (Asset / Name)" string, i guess, but then the annotation wraps information that it's not really annotating - not great.

anyways - some possible alternatives to it being an immutable part of most headers, that may or may not require changes to SourceData:
* it could be a `<span>` of smaller text prepended to the message body, rendered only when it's necessary to distinguish e.g. an asset meter
* it could be always present, but in a place where it's possible to hide it when it's not relevant, e.g. a tooltip
* an indirect solution, or a least something that could somewhat mitigate: i update the labels that dataforged provides. most of the game refers to it as "+companion health" rather than "+health", so there's a case to be made for it being that instead. "integrity" has several canonical permutations: "+command vehicle integrity", etc.
  * not a perfect solution. multiple companions would be more difficult to distinguish

ben commented on src/module/rolls/chat-message.ts:
```diff
@@ -132,8 +236,7 @@ export class IronswornRollChatMessage {
-      this.roll.postAdjustmentOutcome?.value ??
-      this.roll.preAdjustmentOutcome?.value
+      this.roll.finalOutcome?.value ?? this.roll.rawOutcome?.value
     if (!theOutcome) return {}
```
Two downsides of using numbers to encode the roll outcome:

1. In Javascript, 0 is falsy. Currently, any move-roll that misses has no move-outcome text.
2. It's harder to read in the debugger:

![CleanShot 2022-08-23 at 08 04 47](https://user-images.githubusercontent.com/39902/186193987-d5db5a95-68a5-4cf7-989f-3cc947ad429f.jpg)

rsek replied:
oh, good catch! i've added some stricter typing/parsing of numbers/booleans from dialogue options that remedies that.

TBH, i think the `source` being "Roll" there might be a more effective target for revision. in a sense, most things in a roll are sourced from a roll, so it might not be the most effective way of disambiguating the term. further muddying the waters: in the game's parlance, it's still a 'roll' (as in, the mechanical operation that you resolve to an outcome) when the challenge dice come into play, even if you're not actually randomizing any of the dice by rolling them (because they're pre-set, for instance).

so i'd get more specific -- maybe something like "Natural roll outcome" or "Unmodified roll outcome" -- which gives a stronger hint that it's supposed to be a RollOutcome rather than accidentally implying it's a dice value or something. though, the best user-facing label would probably just be "Move outcome" (differs from internal terminology, but it's a more accurate term as far as the game rules are concerned, and that should be a priority for user-facing text), unless the outcome has been modified in some way. otherwise the additional info is just confusing.

so whatever you call it, that item's title might be better handled by the HBS template alone, TBH (and, yes, this is exactly the sort of thing i meant when i expressed misgivings about conflating user-facing labels and internal labels into the same string field ;) )

there's some other implications it has on roll output, but i'll write those up and put them in #366

rsek commented:
i'll mark this as ready for review -- i've got this nagging feeling i've missed something, but i'm not sure what it is yet. i'll do some more thorough testing of it in the mean time.
